CONTRIBUTING.md

```diff
@@ -71,6 +75,72 @@ IntelliJ's code generation can be configured to do this automatically:
      Tick:
      - Make generated local variables final
      - Make generated parameters final
+     
+**Null checking of method or function parameters**
+
+If arguments to a constructor set members of the object, or if arguments to a method
+change internal state, e.g. adding an item to a list, they should be checked for null, e.g.
+
+```
+public class MyObject {
+
+    private final String name;
+    private final List<String> myList = new ArrayList<>();
+
+    public MyObject(final String name) {
+        this.name = Objects.requireNonNull(name, "name");
+    }
+
+    public void addItem(final String item) {
+        myList.add(Objects.requireNotNull(item, "item"));
+    }
+}
+
+```
+
+**No TODOs in code**
+
+Please don't leave any TODOs in the code
+
+**New line for each method or constructor parameter**
+
+Methods or constructors should be defined with each parameter on a new line.
+
+```
+public void doSomething(
+    final String name,
+    final int age,
+    final boolean human
+) {
+    // body of method
+}
+
+```
+
+**Private static methods**
+
+Private methods that don't use object state should be declared as static.
+
+**No else after if branch that returns**
```

Comment:
> Thanks for getting this out there @purplefox! Not sure if we want to discuss the rules here or somewhere else (I actually kinda prefer it on a PR than as an issue because we can have different "threads" centralized in one place) - but let me know.

My original intention was to just catalogue the rules here, hence I haven't given my opinion on whether I think they add value or not, but I am happy to open up the discussion to whether we want these rules or not, if others are?

Reply:
🔥 🔥 🔥 let's do it 🥊 🥊 🥊

Reply:
Ok

Reply:
I'm not a fan of mandating this pattern either (although it can be appropriate in some cases). I think often it can make the intent of method (especially small ones) less clear

Reply:
For me, I'm not so bothered. It is purely a style thing, where as the ones above are more about being able to reason about the code and cod correctness.

I may request someone to change this in the code as I believe it will be more readable after the change, but I'd not block a review on it. 

I generally only flag this up if its used excessively, to the detriment of the code as a whole. Excessive nesting can make the code harder to read and leads to ugly code when we start to hit the line limit.

As an example, I think most people would find this:

```java
if (something) {
   return;
} else {
   if (!somethingElse) {
      bar();
      return;
   } else {
      foo();
      if (anotherThing()) {
         return;
      } else {
         doo();
      }
   }
}
```

Easier to read/scan if it was:

```java
if (something) {
   return;
} 

if (!somethingElse) {
   bar();
   return;
}

foo();

if (anotherThing()) {
   return;
}

doo();
}
```

As for find bugs.... I don't think it will report a higher complexity for:

```java
if (x) {
   y();
   return;
} else {
   z();
   return;
}
```

vs

```java
if (x) {
   y();
   return;
}

 z();
}
```

Though I've not checked.

TBH, if others want, then I'm happy with this not being in the style rules. Instead, we should leave reviewers to make suggestions where they think code can be better laid out, either due to 'if/return/else' or for other reasons.  Though I'm interested to hear what others think.

Reply:
+1 I think your comment makes sense (not part of the rules so up to the implementer, but it's a totally valid suggestion a reviewer can make if they think it'll improve readability)

Reply:
This is the common Java convention, but I wouldn't be surprised if this changes as Java transitions to be more oriented towards expressions instead of statements. Expression based languages favor if/else. Once you get used to the latter, it's hard go back. Anyway, I this was meant more as a note than a recommendation/suggestion.